xcube_stac/impl.py

```diff
@@ -0,0 +1,511 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import json
+from typing import Iterator, Union
+
+import odc.stac
+import odc.geo
+import pystac
+import requests
+import xarray as xr
+from xcube.core.mldataset import MultiLevelDataset
+from xcube.core.store import DataStoreError, DataTypeLike, new_data_store
+from xcube.util.jsonschema import JsonObjectSchema
+
+from .accessor import HttpsDataAccessor, S3DataAccessor
+from .constants import (
+    FloatInt,
+    LOG,
+    MLDATASET_FORMATS,
+    STAC_OPEN_PARAMETERS,
+    STAC_OPEN_PARAMETERS_STACK_MODE,
+    STAC_SEARCH_PARAMETERS,
+    STAC_SEARCH_PARAMETERS_STACK_MODE,
+)
+from .href_parse import _decode_href
+from .mldataset import SingleItemMultiLevelDataset, StackModeMultiLevelDataset
+from .utils import (
+    _apply_scaling_nodata,
+    _convert_datetime2str,
+    _extract_params_xcube_server_asset,
+    _get_data_id_from_pystac_object,
+    _get_format_from_asset,
+    _get_formats_from_item,
+    _get_url_from_pystac_object,
+    _is_xcube_server_asset,
+    _is_valid_ml_data_type,
+    _list_assets_from_item,
+    _search_collections,
+    _search_nonsearchable_catalog,
+    _select_xcube_server_asset,
+    _update_dict,
+    _xarray_rename_vars,
+)
+
+_HTTPS_STORE = new_data_store("https")
+_OPEN_DATA_PARAMETERS = {
+    "open_params_dataset_netcdf": _HTTPS_STORE.get_open_data_params_schema(
+        opener_id="dataset:netcdf:https"
+    ),
+    "open_params_dataset_zarr": _HTTPS_STORE.get_open_data_params_schema(
+        opener_id="dataset:zarr:https"
+    ),
+    "open_params_dataset_geotiff": _HTTPS_STORE.get_open_data_params_schema(
+        opener_id="dataset:geotiff:https"
+    ),
+    "open_params_mldataset_geotiff": _HTTPS_STORE.get_open_data_params_schema(
+        opener_id="mldataset:geotiff:https"
+    ),
+    "open_params_dataset_levels": _HTTPS_STORE.get_open_data_params_schema(
+        opener_id="dataset:levels:https"
+    ),
+    "open_params_mldataset_levels": _HTTPS_STORE.get_open_data_params_schema(
+        opener_id="mldataset:levels:https"
+    ),
+}
+
+
+class SingleImplementation:
```

Comment:
_Single Implementation_ of what? I recommend finding a better name of what is implemented here, for example, the _Store Mode_. Then `StoreMode` would describe the purely abstract interface, and `SingleStoreMode` represents its implementation for the mode _Single_. 

```suggestion
class SingleStoreMode:
```

Reply:
Changed the name to `SingleStoreMode`. To be consistent I also changed to name from `StackImplementation` to `StackStoreMode`.